### PR TITLE
ADD stock_picking_back2draft

### PR DESCRIPTION
**Pull request by eLBati**
This module allows to bring canceled pickings back to draft
Just click on 'back to draft' button

### CODE REVIEW

**Comment by pedrobaeza** [comment]
For consistency with _sale_order_back2draft_ module, we can name this one the same way, don't you think?

**Comment by eLBati** [comment]
@pedrobaeza thanks, done

File: stock_picking_back2draft/__init__.py [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# -*- coding: utf-8 -*-
+# © <2016> Lorenzo Battistini - Agile Business Group
```
Comment (tafaRU) [inline stock_picking_back2draft/__init__.py]: Shouldn't it be only 2016?

File: stock_picking_back2draft/README.rst [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+======================
+Pickings back to draft
+======================
+
+This module allows to bring cancelled pickings back to draft
+
+Usage
+=====
+
+Just click on 'back to draft' button
```
Comment (pedrobaeza) [inline stock_picking_back2draft/README.rst]: Select a canceled picking and click...

File: stock_picking_back2draft/README.rst [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/154/8.0
+
+Bug Tracker
+===========
+
+Bugs are tracked on `GitHub Issues
+<https://github.com/OCA/stock-logistics-workflow/issues>`_. In case of trouble, please
```
Comment (pedrobaeza) [inline stock_picking_back2draft/README.rst:24]: You can include here the link for new issue.

**Comment by eLBati** [comment]
@tafaRU @pedrobaeza thanks, changed

File: stock_picking_back2draft/README.rst (hunk 3 above)
Comment (tafaRU) [inline stock_picking_back2draft/README.rst:24]: @eLBati, the link is still missing!

Comment (eLBati) [inline stock_picking_back2draft/README.rst:24]: I followed the template at https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst
Is it missing some part?

Comment (tafaRU) [inline stock_picking_back2draft/README.rst:24]: @eLBati, I've just seen README file is changed, see https://github.com/OCA/maintainer-tools/commit/6f05e79508e2a4b20f05877ddfd366a59c5c7238 as reference, it's right then :wink:

**Comment by tafaRU** [comment]
:+1:

File: stock_picking_back2draft/models/stock.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+from openerp.exceptions import Warning as UserError
+from openerp.tools.translate import _
+
+
+class StockMove(models.Model):
+    _inherit = 'stock.move'
+
+    @api.multi
+    def action_back_to_draft(self):
+        state_list = self.mapped('state')
```
Comment (pedrobaeza) [inline stock_picking_back2draft/models/stock.py]: You can do:

`if any(lambda x: x.state != 'cancel' for x in self.mapped('state')):`

File: stock_picking_back2draft/models/stock.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+        if set(state_list) != set(['cancel']):
+            raise UserError(_("you can set to draft cancelled moves only"))
```
Comment (pedrobaeza) [inline stock_picking_back2draft/models/stock.py]: s/you/You

File: stock_picking_back2draft/models/stock.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+        self.write({'state': 'draft'})
+
+
+class StockPicking(models.Model):
+    _inherit = 'stock.picking'
+
+    @api.multi
+    def action_back_to_draft(self):
+        state_list = self.mapped('state')
+        if set(state_list) != set(['cancel']):
```
Comment (pedrobaeza) [inline stock_picking_back2draft/models/stock.py]: The same

**Comment by eLBati** [comment]
@pedrobaeza thanks, improved

File: stock_picking_back2draft/README.rst (hunk 3 above)
Comment (pedrobaeza) [inline stock_picking_back2draft/README.rst:24]: Yeah, indeed!

**Comment by pedrobaeza** [comment]
@eLBati, have you seen what happened with the procurements associated to the picking? I think they're canceled (or maybe in exception). They should be rerun.

**Comment by eLBati** [comment]
@pedrobaeza you brought me to mind the need of a link module between `stock_picking_back2draft` and `sale`, see https://github.com/OCA/stock-logistics-workflow/pull/222/commits/12d82d69b80b13cd1301df9a576fe860b88b8e31

About procurements, the one that created the move becomes cancelled after the move is cancelled. It remains cancelled until the move becomes done, then the procurement becomes done too.
I don't see a problem here, as the procurement already generated our move.

Procurements generated by our move are not directly linked to our move, so they should not be affected by cancellation.

Am I missing something? What problems do you see?

**Comment by pedrobaeza** [comment]
@eLBati, thanks for the investigation.

Now it looks good! :+1:

**Comment by andhit-r** [comment]
Functionality tested on runbot :+1: Though it would be nice to have unittest for interaction with procurement order.

File: sale_stock_picking_back2draft/__openerp__.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    "author": "Agile Business Group, Odoo Community Association (OCA)",
+    "license": "AGPL-3",
+    "application": False,
+    "installable": True,
+    'auto_install': True,
+    "depends": [
+        "sale",
+        "stock_picking_back2draft",
+    ],
+    "data": [
```
Comment (tafaRU) [inline sale_stock_picking_back2draft/__openerp__.py]: Please remove this empty key.

**Comment by eLBati** [comment]
@tafaRU thanks, done

**Pull request merged**